src/Microsoft.OData.Core/UriParser/Parsers/CountSegmentParser.cs

```diff
@@ -0,0 +1,138 @@
+        /// <param name="countedInstance">The instance to count on.</param>
+        public CountSegmentToken CreateCountSegmentToken(string identifier, QueryToken countedInstance)
+        {
+            QueryToken filterToken = null;
+            QueryToken searchToken = null;
+            string textWithinParenthesis = null;
+
+            // We need to maintain the recursionDepth of the outer $filter query since calling
+            // ParseFilter or ParseSearch below will reset it to 0.
+            int outerRecursiveDepth = this.UriQueryExpressionParser.recursionDepth;
+            ExpressionLexer outerLexer = this.UriQueryExpressionParser.Lexer;
+
+            if (this.lexer.CurrentToken.Kind != ExpressionTokenKind.OpenParen)
+            {
+                return new CountSegmentToken(identifier, countedInstance);
+            }
+            else
+            {
+                textWithinParenthesis = this.lexer.AdvanceThroughExpandOption();
+            }
+
+            // TODO: check for NoDollarQueryOptions
+            if (textWithinParenthesis.StartsWith(ExpressionConstants.QueryOptionFilter, StringComparison.OrdinalIgnoreCase))
+            {
+                string filterQuery = TryGetQueryOption(ExpressionConstants.QueryOptionFilter, textWithinParenthesis);
+                filterToken = filterQuery == null ? null: this.UriQueryExpressionParser.ParseFilter(filterQuery);
+            }
+            else if (textWithinParenthesis.StartsWith(ExpressionConstants.QueryOptionSearch, StringComparison.OrdinalIgnoreCase))
+            {
+                string searchQuery = TryGetQueryOption(ExpressionConstants.QueryOptionSearch, textWithinParenthesis);
+                searchToken = searchQuery == null ? null : this.UriQueryExpressionParser.ParseSearch(searchQuery);
+            }
+            else
+            {
+                // Only a $filter and $search can be inside a $count()
+                throw ParseError(ODataErrorStrings.UriQueryExpressionParser_IllegalQueryOptioninDollarCount());
+            }
+
+            if (this.lexer.CurrentToken.Kind != ExpressionTokenKind.CloseParen)
+            {
+                throw ParseError(ODataErrorStrings.UriQueryExpressionParser_CloseParenExpected(this.lexer.CurrentToken.Position, this.lexer.ExpressionText));
+            }
+
+            this.lexer.NextToken();
+            this.UriQueryExpressionParser.recursionDepth = outerRecursiveDepth;
+            this.UriQueryExpressionParser.Lexer = outerLexer;
+            return new CountSegmentToken(identifier, countedInstance, filterToken, searchToken);
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="queryOptionName"></param>
+        /// <param name="query"></param>
+        /// <returns></returns>
+        private static string TryGetQueryOption(string queryOptionName, string query)

```

Comment:
>TryGetQueryOption [](http://example.com/codeflow?start=30&length=17)

I think this entire method goes away if you first check to see if the query option is $filter/$search, and then use AdvanceThroughExpandOption to just read the value.

Right now it seems like you get the expression for the key value pair, check to see if it is filter or search, and then try to remove the filter/search so that you can pass it to the appropriate Parse method.

Reply:
Resolved